snuba/query/parser/strings.py

```diff
@@ -13,14 +13,22 @@ def parse_string_to_expr(val: str) -> Expression:
     in the Snuba Query.
     """
     # TODO: This will use the schema of the dataset/entity to decide if the expression is
-    # a column or a literal.
+    # a column or a literal. In the meantime, make some guesses about the value.
+
+    if QUOTED_LITERAL_RE.match(val):
+        return Literal(None, val[1:-1])
+
     if val.isdigit():
         return Literal(None, int(val))
-    else:
-        try:
-            return Literal(None, float(val))
-        except Exception:
-            if QUOTED_LITERAL_RE.match(val):
-                return Literal(None, val[1:-1])
-            else:
-                return Column(None, None, val)
+
+    try:
+        return Literal(None, float(val))
+    except Exception:
+        pass
+
+    try:
+        return Literal(None, parse_datetime(val))
```

Comment:
If I remember correctly we do not provide quotes for string conditions. 
`["tags[environment]", "=", "prod"]`
instead
`["tags[environment]", "=", "'prod'"]`

That would mean we would assume a string that properly parses would be considered a datetime while it could be a string that only happens to contain a datetime. This would cause the opposite issue with respect to what we had today. 

I don't think there is a reliable way to disambiguate the two cases without introducing a special syntax for datetime that allows us to definitively tell when something is a string vs datetime.

Reply:
Example:
`["tags[my_weird_time_tag]", =", "2020-01-01 00:00:00"]`
which would turn into
`arrayElement(tags.value, indexOf(tags.key, 'my_weird_time_tag')) = toDateTime('2020-01-01 00:00:00')` which would fail the query since tags.value is a string.

Reply:
Ah no nevermind. The right hand side of a condition is not built through this code. Then this is not going to be aproblem there. 
Could you please verify we do not do the same trick in any other place where we are supposed to generate literals ?